mllib/src/main/scala/org/apache/spark/mllib/evaluation/BinaryClassificationMetrics.scala

```diff
@@ -165,13 +166,17 @@ class BinaryClassificationMetrics @Since("3.0.0") (
     confusions: RDD[(Double, BinaryConfusionMatrix)]) = {
     // Create a bin for each distinct score value, count weighted positives and
     // negatives within each bin, and then sort by score values in descending order.
-    val counts = scoreLabelsWeight.combineByKey(
+    val binnedWeights = scoreLabelsWeight.combineByKey(
       createCombiner = (labelAndWeight: (Double, Double)) =>
         new BinaryLabelCounter(0.0, 0.0) += (labelAndWeight._1, labelAndWeight._2),
       mergeValue = (c: BinaryLabelCounter, labelAndWeight: (Double, Double)) =>
         c += (labelAndWeight._1, labelAndWeight._2),
       mergeCombiners = (c1: BinaryLabelCounter, c2: BinaryLabelCounter) => c1 += c2
-    ).sortByKey(ascending = false)
+    )
+    if (scoreLabelsWeight.getStorageLevel != StorageLevel.NONE) {
+      binnedWeights.persist()
+    }
+    val counts = binnedWeights.sortByKey(ascending = false)
```

Comment:
Wait, hm, I don't understand this. You persist binnedWeights, but it is now only used once. Why? If anything it's binnedCounts that needs persisting. I'm still not clear if it makes enough difference to matter.

Reply:
`binnedCounts` is a child RDD of `binnedWeights`. And here one action `sortByKey` is performed on `binnedWeights`.

Reply:
Yes, but, why bother persisting binnedWeights? you recompute everything in between it and binnedCounts twice, when I think that would be the point, to avoid that.

Reply:
I think `binnedWeights` is required to be persisted because

`binnedWeights`
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;| `sortByKey` (action)
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;V
`counts`
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;| `count` (action)
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;V
`binnedCounts` (on which action `collect` is applied to compute `agg`)

Reply:
I might be wrong here. Kindly correct me @srowen

Reply:
caching helps where more than one action is performed on the same RDD. That's not the case here. Each of the first two has one thing executed on it. sortByKey is not an action, anyway.

Reply:
Oh, okay. One question here, will it be worth persisting `counts` since actions `count` and `collect` is applied directly on it ?

Reply:
Doesn't seem so. But that is the question I'd put to you in these cases - are you sure it makes a difference meaningful enough to overcome the overhead? I could imagine so here, just wondering if these are based on more investigation or benchmarking, vs just trying to persist lots of things.

Reply:
> are you sure it makes a difference meaningful enough to overcome the overhead?

I think, no. Persisting `count` doesn't makes sense here. It will just be an overhead. Now I am getting clear picture of where to use persist. 
Key learnings from this PR about persist.

- persist introduce memory and CPU overheads.

- So only important inputs (such as intermediate results, user data which is already cached, etc) should be persisted or RDD on which more than one action is performed.

- Avoid using persist in loop.

- Persist should be meaningful enough to overcome overheads.

Reply:
TYSM @srowen . Looking for more learning opportunities.